certora/specs/AccessControlDefaultAdminRules.spec

```diff
@@ -0,0 +1,379 @@
+    "Default admin is set to the previous pending default admin";
+  assert success => _pendingDefaultAdmin(e) == 0, 
+    "Pending default admin is reset";
+  assert success => _pendingDefaultAdminSchedule(e) == 0,
+    "Pending default admin delay is reset";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: cancelDefaultAdminTransfer resets pending default admin and its schedule                      │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule cancelDefaultAdminTransfer(env e) {
+  require nonpayable(e);
+  requireInvariant defaultAdminConsistency(defaultAdmin());
+  requireInvariant singleDefaultAdmin(e.msg.sender, defaultAdmin());
+
+  cancelDefaultAdminTransfer@withrevert(e);
+  bool success = !lastReverted;
+
+  // liveness
+  assert success <=> e.msg.sender == defaultAdmin(),
+    "only the current default admin can cancel a transfer";
+
+  // effect
+  assert success => _pendingDefaultAdmin(e) == 0, 
+    "Pending default admin is reset";
+  assert success => _pendingDefaultAdminSchedule(e) == 0,
+    "Pending default admin delay is reset";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: changeDefaultAdminDelay sets a pending default admin delay and its schedule                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule changeDefaultAdminDelay(env e, uint48 newDelay) {
+  require nonpayable(e);
+  require timeSanity(e);
+  require delayChangeWaitSanity(e, newDelay);
+  requireInvariant defaultAdminConsistency(defaultAdmin());
+  requireInvariant singleDefaultAdmin(e.msg.sender, defaultAdmin());
+
+  uint48 delayBefore = defaultAdminDelay(e);
+
+  changeDefaultAdminDelay@withrevert(e, newDelay);
+  bool success = !lastReverted;
+
+  // liveness
+  assert success <=> e.msg.sender == defaultAdmin(), 
+    "only the current default admin can begin a delay change";
+
+  // effect
+  assert success => _pendingDelay(e) == newDelay, "pending delay is set";
+  assert success => (
+    _pendingDelaySchedule(e) > e.block.timestamp || 
+    delayBefore == newDelay || // Interpreted as decreasing, x - x = 0
+    defaultAdminDelayIncreaseWait() == 0
+  ),
+    "pending delay schedule is set in the future unless accepted edge cases";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: pending delay wait is set depending on increasing or decreasing the delay                                     │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pendingDelayWait(env e, uint48 newDelay) {
```

Comment:
This rule is proving our code correct against a reference model, but it would be nice to have a more abstract rule that proves instead _the property that we were aiming for_: that if you change the delay wait and then change the admin, the old delay must have passed in between. Do you see what I mean? I feel it should be possible to do this with two separate `env` variables.

Reply:
Yes, I see what you mean. Will keep playing with this and I'll update if I can come up with an example for that property.

Reply:
Check this out in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4106:

![image](https://user-images.githubusercontent.com/481465/234019291-7f6e234d-96d8-4a2a-b432-39cca1eb6483.png)